ctrd/container_lock.go

```diff
@@ -27,3 +30,27 @@ func (l *containerLock) Unlock(id string) {
 	defer l.mutex.Unlock()
 	delete(l.ids, id)
 }
+
+func (l *containerLock) TrylockWithTimeout(id string, t time.Duration) bool {
+	var retry = 32
+
+	lctx, cancel := context.WithTimeout(context.TODO(), t)
```

Comment:
This is a brand new context? rather than an inherited context?

Reply:
What about using two timer?
```go
func (l *containerLock) TrylockWithTimeout(id string, t time.Duration) bool {
	var retry = 32

	timeoutTimer := time.NewTimer(t)
	for {
		ok := l.Trylock(id)
		if ok {
			timeoutTimer.Stop()
			return true
		}

		// sleep random duration by retry
		select {
		case <-time.After(time.Millisecond * time.Duration(rand.Intn(retry))):
			if retry < 2048 {
				retry = retry << 1
			}
			continue
		case <-timeoutTimer.C:
			return false
		}
	}
}
```

Reply:
I think both are the same.